flight/Modules/VtolPathFollower/vtol_follower_control.c

```diff
@@ -496,22 +389,69 @@ static int32_t vtol_follower_control_accel(float dT)
 	return 0;
 }
 
+
+static float vtol_follower_control_altitude(float downCommand) {
+	AltitudeHoldStateData altitudeHoldState;
+	altitudeHoldState.VelocityDesired = downCommand;
+	altitudeHoldState.Integral = vtol_pids[DOWN_VELOCITY].iAccumulator / 1000.0f;
+	altitudeHoldState.AngleGain = 1.0f;
+
+	if (altitudeHoldSettings.AttitudeComp > 0) {
+		// Throttle desired is at this point the mount desired in the up direction, we can
+		// account for the attitude if desired
+		AttitudeActualData attitudeActual;
+		AttitudeActualGet(&attitudeActual);
+
+		// Project a unit vector pointing up into the body frame and
+		// get the z component
+		float fraction = attitudeActual.q1 * attitudeActual.q1 -
+				 attitudeActual.q2 * attitudeActual.q2 -
+				 attitudeActual.q3 * attitudeActual.q3 +
+				 attitudeActual.q4 * attitudeActual.q4;
+
+		// Add ability to scale up the amount of compensation to achieve
+		// level forward flight
+		fraction = powf(fraction, (float) altitudeHoldSettings.AttitudeComp / 100.0f);
+
+		// Dividing by the fraction remaining in the vertical projection will
+		// attempt to compensate for tilt. This acts like the thrust is linear
+		// with the output which isn't really true. If the fraction is starting
+		// to go negative we are inverted and should shut off throttle
+		downCommand = (fraction > 0.1f) ? (downCommand / fraction) : 0.0f;
+
+		altitudeHoldState.AngleGain = 1.0f / fraction;
+	}
+
+	altitudeHoldState.Throttle = downCommand;
+	AltitudeHoldStateSet(&altitudeHoldState);
+
+	return downCommand;
+}
+
 /**
  * Compute desired attitude from the desired velocity
  * @param[in] dT the time since last evaluation
+ * @param[in] att_adj an adjustment to the attitude for loiter mode
  *
  * Takes in @ref NedActual which has the acceleration in the
  * NED frame as the feedback term and then compares the
  * @ref VelocityActual against the @ref VelocityDesired
  */
-int32_t vtol_follower_control_attitude(float dT)
+int32_t vtol_follower_control_attitude(float dT, const float *att_adj)
 {
 	vtol_follower_control_accel(dT);
 
+	float default_adj[2] = {0,0};
+
+	if (!att_adj) {
+		att_adj = default_adj;
+	}
+
 	AccelDesiredData accelDesired;
 	AccelDesiredGet(&accelDesired);
 
-	StabilizationDesiredData stabDesired;
+	StabilizationSettingsData stabSet;
+	StabilizationSettingsGet(&stabSet);
```

Comment:
it's probably best to move updating this to the settings update method below (make sure you register the callback). It increases the static memory usage (really all the settings should move into a dynamically allocated structure) but avoids additional semaphore locks here.